dilated-file-transfer.md

```diff
@@ -0,0 +1,513 @@
+
+
+## Protocol Details
+
+See the Dilation document for details on the Dilation setup procedure.
+Once a Dilation-supporting connection is open, we will have a "control" subchannel (subchannel #0).
+Either peer can also open additional subchannels.
+
+All control-channel messages are encoded using `msgpack`.
+   --> XXX: see "message encoding" discussion
+
+Control-channel message formats are described using Python pseudo-code to illustrate the data types involved.
+They are actually an encoded `Map` with `String` keys (to use `msgpack` lingo) and values as per the pseudo-code.
+
+All control-channel messages contain a integer "kind" field describing the sort of message it is.
+(That is, `"kind": "text"` for example, not the single-byte tag used for subchannel messages)
+
+Rejected idea: Version message, because we already do version negotiation via mailbox features.
+
+Rejected idea: Offer/Answer messages via the control channel: we need to open a subchannel anyway and the subchannel-IDs are not intended to be part of the Dilation public API.
+
+
+### Control Channel Messages
+
+Each side MAY send a free-form text message at any time.
+These messages look like:
+
+```python
+class Message:
+    message: str     # unicode string
+    kind: str = "text"
+```
+
+
+### Making an Offer
+
+Either side MAY propose any number of Offers at any time after the connection is set up.
+If the other peer specified `"mode": "send"` then this peer MUST NOT make any Offers.
+If this peer specified `"mode": "receive"` then this peer MUST NOT make any Offers.
+
+To make an Offer the peer opens a subchannel.
+Recall from the Dilation specification that subchannels are _record_ pipes (not simple byte-streams).
+
+All records on the subchannel begin with a single byte indicating the kind of message.
+Any additional bytes are a kind-dependent payload.
+
+The following kinds of messages exist (as indicated by the first byte):
+* 1: msgpack-encoded `FileOffer` message
+* 2: msgpack-encoded `DirectoryOffer` message
+* 3: msgpack-encoded `OfferAccept` message
+* 4: msgpack-encoded `OfferReject` message
+* 5: raw file data bytes
+* 6: msgpack-encoded `FileAcknowledge` message
+
+All other byte values are reserved for future use and MUST NOT be used.
+
+    XXX: maybe spec [0, 128) as reserved, and [128, 255) for "experiments"?
+
+The sender that opened the new subchannel MUST immediately send one of the two kinds of offer messages.
+
+To offer a single file (with message kind `1`):
+
+```python
+class FileOffer:
+    filename: str   # filename (no path segments)
+    timestamp: int  # Unix timestamp (seconds since the epoch in GMT)
+    bytes: int      # total number of bytes in the file
+```
+
+At the end of the file data, _both_ sides MUST send a `FileAcknowledge` message:
+
+```python
+class FileAcknowledge:
+    bytes: int   # total number of bytes in the file
+    hash: bytes  # 32-byte Blake2b hash of the file
```

Comment:
Bikeshed: Why Blake2b as hash algorithm and not sha-2 (already used in Magic Wormhole, no new dependencies) or sha-3 (NIST standardized)?

Reply:
I guess actually blake2s (for some reason, presumably to do with "be fast on not-64-bit systems").

We could bikeshed further, but I would vote for Blake2 because:

- not NIST
- fast
- already used in Dilation (via the Noise protocol selection)
- I know one of the authors

There will be a (possibly transient) dependency on some Blake2 and ChaChaPoly1305 library in any case due to the Noise variant selected by Dilation (which is `Noise_NNpsk0_25519_ChaChaPoly_BLAKE2s`).